appinventor/buildserver/src/com/google/appinventor/buildserver/FormPropertiesAnalyzer.java

```diff
@@ -169,21 +169,31 @@ private static void collectComponentTypes(JSONObject componentProperties,
       DocumentBuilder builder = factory.newDocumentBuilder();
       Document doc = builder.parse(new ByteArrayInputStream(source.getBytes(StandardCharsets.UTF_8)));
       NodeList blocks = doc.getElementsByTagName("block");
-      for (int i = 0; i < blocks.getLength(); i++) {
+      Node parent;
+      FOR: for (int i = 0; i < blocks.getLength(); i++) {
         Element block = (Element) blocks.item(i);
         String type = block.getAttribute("type");
 
         // Skip non-component and disabled blocks
         if (!isComponentBlock(type) || "true".equals(block.getAttribute("disabled"))) continue;
-
+        parent = block;
```

Comment:
Rather than do this here, what I think would be more appropriate would be to make a new `isBlockDisabled` function that checks the block's disabled status AND puts the information into a map to memoize it. As implemented, this is O(n*m) in the depth of the tree and number of leaf nodes, rather than O(n) (every ancestor is checked by all of its descendants). This would also prevent having to jump out of the inner loop to the outer loop.